daceml/autodiff/library/library.py

```diff
@@ -0,0 +1,351 @@
+"""
+Dace library for autodiff
+
+Includes the BackwardPass library node, and the replacements for the python frontend
+"""
+from typing import Dict, Set, Optional
+import copy
+
+import dace
+import dace.library
+from dace import SDFG, SDFGState, nodes, data, properties
+from dace.transformation import transformation as pm
+from dace.transformation.passes import analysis
+from dace.sdfg import graph
+
+from daceml.autodiff import backward_pass_generator as engine, analysis as autodiff_analysis
+from daceml.util.utils import in_edge_with_name
+
+
+@properties.make_properties
+class Array(data.Array):
+    """
+    A array for which a gradient can be computed.
+
+    This also has to have the name 'Array' because otherwise none of the python frontend replacements work.
+    """
+    # since this can be None, this is not a DataProperty
+    gradient = properties.Property(dtype=str,
+                                   desc="The corresponding gradient buffer",
+                                   optional=True,
+                                   default=None)
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+    def __repr__(self):
+        return "Parameter" + data.Array.__repr__(self)
+
+    def add_gradient_buffer(self, sdfg: SDFG, name: str) -> str:
+        """
+        Find or create a gradient buffer for the parameter in the given SDFG.
+
+        :param sdfg: the SDFG containing the parameter
+        :param name: the name of the parameter
+        :return: the name of the gradient buffer
+        """
+
+        if self.gradient:
+            return self.gradient
+
+        # First, check if this array already has a gradient buffer a nested SDFG
+
+        cands = set()
+        for state in sdfg.nodes():
+            for node in state.nodes():
+                if not isinstance(node, nodes.NestedSDFG):
+                    continue
+
+                nested_names = set()
+
+                for edge in state.in_edges(node):
+                    if edge.data.data == name:
+                        nested_names.add(edge.dst_conn)
+                for edge in state.out_edges(node):
+                    if edge.data.data == name:
+                        nested_names.add(edge.dst_conn)
+
+                for name in nested_names:
+                    nested_desc = node.sdfg.arrays[name]
```

Comment:
This only goes one level deep. Explain why this is necessary or if we need to traverse further
Also, why do you take the internal descriptor rather than the external one, which is the one that may fit the array better (e.g., grad of A[:, 1])

Reply:
Here's the comment I will add, let me know if this clears it up:
```
 First, check if this array already has a gradient buffer in a nested SDFG
 This happens, for example when pytorch modules are used in the frontend. In that case:
 1. the parser assembles the closure of the module, which adds
    descriptors for all the parameters and their gradients (if they are
    required).
 2. The DaceProgram will then pass these arrays in when the
     DaceProgram is called, using the names from the inner closure.
 3. When parsing the backward nodes, we want the gradient buffers in
    the closure to match the gradient buffers that we pass in. Thus, we
    need to make sure that we use the same name as the NestedSDFG

 Note that we do not currently do any nesting beyond this level,
 because nested modules are converted to one SDFG.
```